src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1252,5 +1245,236 @@ internal bool CanAddParsedValue(HttpHeaderParser parser)
+            }
+
+            ref object? GrowEntriesAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                if (entries.Length == ArrayThreshold)
+                {
+                    return ref ConvertToDictionaryAndAddDefault(key);
+                }
+                else
+                {
+                    Array.Resize(ref entries, entries.Length << 1);
+                    _headerStore = entries;
+                    ref HeaderEntry firstNewEntry = ref entries[entries.Length >> 1];
+                    firstNewEntry.Key = key;
+                    return ref firstNewEntry.Value!;
+                }
+            }
+
+            ref object? ConvertToDictionaryAndAddDefault(HeaderDescriptor key)
+            {
+                var entries = (HeaderEntry[])_headerStore!;
+                var dictionary = new Dictionary<HeaderDescriptor, object>(ArrayThreshold);
+                _headerStore = dictionary;
+                foreach (HeaderEntry entry in entries)
+                {
+                    dictionary.Add(entry.Key, entry.Value);
+                }
+                Debug.Assert(dictionary.Count == _count - 1);
+                return ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+            }
+
+            ref object? DictionaryGetValueRefOrAddDefault(HeaderDescriptor key)
+            {
+                var dictionary = (Dictionary<HeaderDescriptor, object>)_headerStore!;
+                ref object? value = ref CollectionsMarshal.GetValueRefOrAddDefault(dictionary, key, out s_dictionaryGetValueRefOrAddDefaultExistsDummy);
+                if (value is null)
+                {
+                    _count++;
+                }
+                return ref value;
+            }
+        }
+
+        private void AddEntryToStore(HeaderEntry entry)
+        {
+            Debug.Assert(!ContainsKey(entry.Key));
+
+            if (_headerStore is HeaderEntry[] entries)
+            {
+                int count = _count;
+                if ((uint)count < (uint)entries.Length)
+                {
+                    entries[count] = entry;
+                    _count++;
+                    return;
+                }
+            }
+
+            GetValueRefOrAddDefault(entry.Key) = entry.Value;
+        }
+
+        internal bool ContainsKey(HeaderDescriptor key)
+        {
+            return !Unsafe.IsNullRef(ref GetValueRefOrNullRef(key));
+        }
+
+        public void Clear()
+        {
+            object? store = _headerStore;
+            if (store is not null)
+            {
+                if (store is HeaderEntry[] entries)
+                {
+                    Array.Clear(entries, 0, _count);
+                }
+                else
+                {
+                    Unsafe.As<Dictionary<HeaderDescriptor, object>>(store).Clear();
```

Comment:
Presumably we're not reverting back to the array-based strategy on purpose?

Reply:
Given how edge-case having the dictionary + calling `Headers.Clear()` is, I didn't think it matters much either way.
I suppose that if someone is brave enough to reuse the request and headers, we shouldn't force them into the slow path if a single request had a lot of headers. I'll change this to just `null` out the field.